### PR TITLE
feat: introduce yarn workspace scanning test & monitor only

### PR DESCRIPTION
**Pull request by lili2311**
- [x] Ready for review
- [x] Follows [CONTRIBUTING](https://github.com/snyk/snyk/blob/master/.github/CONTRIBUTING.md) rules
- [x] Reviewed by Snyk internal team

To do:
- [x] Add a test with policies nested
- [x] Fix policies handling | [PR1](https://github.com/snyk/snyk/pull/1223), [PR2](https://github.com/snyk/snyk/pull/1224)
- [x] Specific error if no suported manifests found
- [x] Scanning multiple yarn workspaces 

#### What does this PR do?
Introduce yarn-workspaces CLI support by allowing a user to start the test with `snyk test --yarn-workspaces`
Similar to all projects it can be tailored to go X levels deep in the repo.

For all files that are found to be match the `workspaces` pattern they will be processed with the root lockfile as a reference.

Support `test` & `monitor` only. No specific support for `wizard` or `patch`

#### Where should the reviewer start?
https://github.com/snyk/snyk/compare/feat/yarn-workspaces?expand=1#diff-7caec6ff8dcde1d56167613fa8d3ca0cR185

#### How should this be manually tested?

Example repo: https://github.com/snyk-fixtures/yarn-workspace-octan-lite

#### Any background context you want to provide?
Adding dedicated support for `yarn workspaces` where a lockfile lives in a root of multiple sub-projects with a root projects. Previously assumed support always expected a lockfiel or node_modules next to a manifest which is completely different for workspaces.

#### What are the relevant tickets?
https://github.com/snyk/snyk/issues/350
#### Screenshots
ezgif com-video-to-gif (1)

### CODE REVIEW

File: src/lib/plugins/nodejs-plugin/yarn-workspaces-parser.ts [hunk 1]
```diff
@@ -0,0 +1,141 @@
...
+    for (const workspaceRoot of Object.keys(yarnWorkspacesMap)) {
+      const workspaces = yarnWorkspacesMap[workspaceRoot].workspaces || [];
+      const match = workspaces
+        .map((pattern) =>
+          packageJsonFileName.includes(pattern.replace(/\*/, '')),
+        )
+        .filter(Boolean);
+
+      if (match) {
+        _.set(
```
Comment (dkontorovskyy) [inline src/lib/plugins/nodejs-plugin/yarn-workspaces-parser.ts:52]: `{... yarnWorkspacesFilesMap, { [packageJsonFileName]: {root: workspaceRoot} }};`
This should work. Let's not use lodash unless we 100% need it

File: src/lib/plugins/nodejs-plugin/yarn-workspaces-parser.ts [hunk 2]
```diff
@@ -0,0 +1,141 @@
...
+    return yarnWorkspacesMap;
+  }
+
+  try {
+    const rootFileWorkspacesDefinitions = lockFileParser.getYarnWorkspaces(
+      file.content,
+    );
+
+    if (rootFileWorkspacesDefinitions && rootFileWorkspacesDefinitions.length) {
+      _.set(
```
Comment (dkontorovskyy) [inline src/lib/plugins/nodejs-plugin/yarn-workspaces-parser.ts:131]: same here about removing lodash

File: test/acceptance/cli-monitor/cli-monitor.all-projects.spec.ts [hunk 3]
```diff
@@ -12,6 +12,116 @@ interface AcceptanceTests {
...
+      const loadPlugin = sinon.stub(params.plugins, 'loadPlugin');
+      t.teardown(loadPlugin.restore);
+
+      const result = await params.cli.monitor('yarn-workspaces', {
+        yarnWorkspaces: true,
+        detectionDepth: 5,
+      });
+      t.ok(loadPlugin.withArgs('npm').notCalled, 'skips load plugin');
+
+      console.log('*** output', result);
```
Comment (dkontorovskyy) [inline test/acceptance/cli-monitor/cli-monitor.all-projects.spec.ts:78]: console.log leftover? :)

File: src/lib/options-validator.ts [hunk 4]
```diff
@@ -9,7 +9,7 @@ export async function validateOptions(
 ): Promise<void> {
   if (options.reachableVulns) {
     // Throwing error only in case when both packageManager and allProjects not defined
```
Comment (orsagie) [inline src/lib/options-validator.ts:11]: amend comment or throw different error

File: src/lib/plugins/get-deps-from-plugin.ts [hunk 5]
```diff
@@ -13,23 +13,36 @@ import {
 import analytics = require('../analytics');
 import { convertSingleResultToMultiCustom } from './convert-single-splugin-res-to-multi-custom';
 import { convertMultiResultToMultiCustom } from './convert-multi-plugin-res-to-multi-custom';
+import { processYarnWorkspaces } from './nodejs-plugin/yarn-workspaces-parser';
 
 const debug = debugModule('snyk-test');
 
+const processors = {
```
Comment (orsagie) [inline src/lib/plugins/get-deps-from-plugin.ts:20]:
```suggestion
const multiProjectProcessors?? = {
```

File: src/lib/plugins/get-deps-from-plugin.ts [hunk 6]
```diff
@@ -13,23 +13,36 @@ import {
...
+
 // Force getDepsFromPlugin to return scannedProjects for processing
 export async function getDepsFromPlugin(
   root: string,
   options: Options & (TestOptions | MonitorOptions),
 ): Promise<pluginApi.MultiProjectResult> {
   let inspectRes: pluginApi.InspectResult;
 
-  if (options.allProjects) {
+  if (options.yarnWorkspaces || options.allProjects) {
```
Comment (orsagie) [inline src/lib/plugins/get-deps-from-plugin.ts:38]: I want to have something like the following to be clear that this is the path for multiple projects, but it doesn't look quite right to me
```suggestion
  if (options.keys().some(multiProjectProcessors.keys())
```

File: src/cli/commands/monitor/formatters/format-monitor-response.ts [hunk 7]
```diff
@@ -22,9 +22,6 @@ export function formatMonitorOutput(
     : res.path;
   const strOutput =
     chalk.bold.white('\nMonitoring ' + humanReadableName + '...\n\n') +
-    (packageManager === 'yarn'
```
Comment (lili2311) [inline src/cli/commands/monitor/formatters/format-monitor-response.ts:25]: seems to be some old leftover message, not useful

File: help/help.txt [hunk 8]
```diff
@@ -121,6 +121,8 @@ npm options:
 Yarn options:
   --strict-out-of-sync=<true|false>
                        Prevent testing out of sync lockfiles. Defaults to true.
+  --yarn-workspaces    Detect and scan yarn workspaces. Similar to --all-projects you can specify how
+                       many sub-directories to search & exclude some directories.
```
Comment (gitphill) [inline help/help.txt:125]: you can specify how many sub-directories to search using --detection-depth and exclude directories using --exclude.

File: src/cli/index.ts [hunk 9]
```diff
@@ -241,18 +241,61 @@ async function main() {
...
+        'yarn-workspaces',
+      ]);
+    }
+
+    if (args.options.allSubProjects && args.options.yarnWorkspaces) {
+      throw new UnsupportedOptionCombinationError([
+        'all-sub-projects',
+        'yarn-workspaces',
+      ]);
+    }
```
Comment (gitphill) [inline src/cli/index.ts:297]: is there a nice way of doing this in a loop over map of unsupported combinations?

Comment (lili2311) [reply src/cli/index.ts]: pushed changes to refactor this

**Comment by lili2311** [comment]
> You got some errors on this test, updating its logic with your new changes might solve the problem
> 
> 1. test/acceptance/cli-test/cli-test.acceptance.test.ts Languages Yarn test --yarn-workspaces --detection-depth=5 multiple workspaces found Testing /home/circleci/snyk/test/acceptance/workspaces...  CustomError: Could not detect supported target files in

not seeing any failure on this one, looks good to me?

**Comment by lili2311** [comment]
Now I see it, looking.

File: src/lib/errors/exclude-flag-invalid-input.ts [hunk 10]
```diff
@@ -3,7 +3,7 @@ import { CustomError } from './custom-error';
 export class ExcludeFlagInvalidInputError extends CustomError {
   private static ERROR_CODE = 422;
   private static ERROR_MESSAGE =
-    'The --exclude argument must be a comma seperated list of directory names and cannot contain a path.';
+    'The --exclude argument must be a comma separated list of directory names and cannot contain a path.';
```
Comment (lili2311) [inline src/lib/errors/exclude-flag-invalid-input.ts:6]: typo

**Comment by lili2311** [comment]
@anthogez @gitphill  the test that tries to test multiple workspaces fails because if we do not find any manifest in first level we stop and throw. This seems to be intended behaviour so I am going to add an empty package.json to the root to show the multi-workspace test working rather than copy those folders into a new one for the sake of a test
```
  snyk-detect no file specified. Trying to autodetect in base folder /Users/lili/www/snyk/snyk/test/acceptance/workspaces +0ms
  snyk-detect no package file found in /Users/lili/www/snyk/snyk/test/acceptance/workspaces +0ms
```

**Pull request merged**